docs/mitigating_ret2spec.md

```diff
@@ -0,0 +1,122 @@
+# Mitigating ret2spec
+
+This doc explains ret2spec attacks and how they are mitigated between different execution modes. In particular, we explain the subtleties of how the Linux kernel is protected from ret2spec attacks originating from user processes.
+
+## Background: anatomy of a ret2spec attack
+
+Function calls on x86 are built on two instructions: `CALL` and `RET`. `CALL` pushes the address of the next instruction -- the _return address_ -- onto the stack, then jumps to the start of the function being called. When that function is done, it uses `RET` to pop the return address off the stack and jump back to it.
+
+It can take a while for a `RET` to fully resolve the return address -- for example, if the address has been evicted from cache. To avoid stalling out-of-order execution, modern processors use a _return stack buffer_ (RSB) to remember return addresses. When a `CALL` executes, the return address is pushed onto the RSB; later, when we get to a `RET`, the top entry is popped off the RSB and used as the predicted return address.
```

Comment:
(re RSB): nit, but new terms should be bold rather than italic.

There's a few places where this happens, I can write the follow-up PR if you like,

Reply:
this one seems like a question of taste -- the styles I'm familiar with use italics. See e.g. https://blog.apastyle.org/apastyle/2015/04/using-italics-for-technical-or-key-terms.html